apis/core/lib/domain/csv-source/upload.ts

```diff
@@ -1,46 +1,52 @@
 import { GraphPointer } from 'clownface'
 import { schema } from '@tpluscode/rdf-ns-builders'
+import type { MediaObject } from '@rdfine/schema'
 import { Conflict } from 'http-errors'
 import { NamedNode } from 'rdf-js'
-import * as s3 from '../../storage/s3'
 import { error } from '../../log'
 import { ResourceStore } from '../../ResourceStore'
 import { loadFileHeadString } from '../csv/file-head'
 import { sniffParse } from '../csv'
 import { sampleValues } from '../csv/sample-values'
 import * as CsvSourceQueries from '../queries/csv-source'
 import { CsvMapping } from '@cube-creator/model'
+import { cc } from '@cube-creator/core/namespace'
+import type { MediaStorage } from '../../storage'
+import { getMediaStorage } from '../../storage'
 
 interface CreateCSVSourceCommand {
   csvMappingId: NamedNode
   resource: GraphPointer
   store: ResourceStore
-  fileStorage?: s3.FileStorage
+  getStorage?: (m: MediaObject) => MediaStorage
   csvSourceQueries?: Pick<typeof CsvSourceQueries, 'sourceWithFilenameExists'>
 }
 
 export async function createCSVSource({
   csvMappingId,
   resource,
   store,
-  fileStorage = s3,
+  getStorage = getMediaStorage,
   csvSourceQueries: { sourceWithFilenameExists } = CsvSourceQueries,
 }: CreateCSVSourceCommand): Promise<GraphPointer> {
   const csvMapping = await store.getResource<CsvMapping>(csvMappingId)
 
+  const sourceKind = resource.out(cc.sourceKind).term! as NamedNode
   const fileName = resource.out(schema.name).value!
-  const key = resource.out(schema.identifier).value!
+  const key = resource.out(schema.identifier).value! || ''
```

Comment:
Why is it ok for the identifier to be missing?

Also:
```suggestion
  const key = resource.out(schema.identifier).value || ''
```

Reply:
Because identifier is only used for files stored in s3. But I can come up with a value for other files as well if you think it makes sense.